go/vt/vttablet/tabletserver/semisyncwatcher/watcher.go

```diff
@@ -0,0 +1,282 @@
+		waiters:    make([]chan any, 0),
+	}
+}
+
+// Open starts the watcher.
+func (w *Watcher) Open() {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	if w.isOpen {
+		// If we are already open, then there is nothing to do
+		return
+	}
+	// Set the watcher to be open.
+	w.isOpen = true
+	log.Info("SemiSync Watcher: opening")
+
+	// This function could be running from within a unit test scope, in which case we use
+	// mock pools that are already open. This is why we test for the pool being open.
+	if !w.appPool.IsOpen() {
+		w.appPool.Open(w.env.Config().DB.AppWithDB())
+	}
+	w.clearTicks.Start(w.clearAllData)
+	w.ticks.Start(w.checkAndFixSemiSyncBlocked)
+}
+
+// Close stops the watcher.
+func (w *Watcher) Close() {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	if !w.isOpen {
+		// If we are already closed, then there is nothing to do
+		return
+	}
+	w.isOpen = false
+	log.Info("SemiSync Watcher: closing")
+	w.clearTicks.Stop()
+	w.ticks.Stop()
+	w.appPool.Close()
+}
+
+// checkAndFixSemiSyncBlocked checks if the primary is blocked on semi-sync ack
+// and manufactures a write to unblock the primary. This function is safe to
+// be called multiple times in parallel.
+func (w *Watcher) checkAndFixSemiSyncBlocked() {
+	// Check if semi-sync is blocked or not
+	isBlocked, err := w.isSemiSyncBlocked(context.Background())
+	if err != nil {
+		// If we are unable to determine whether the primary is blocked or not,
+		// then we can just abort the function and try again later.
+		log.Errorf("SemiSync Watcher: failed to check if primary is blocked on semi-sync: %v", err)
+		return
+	}
+	// Set the isBlocked state.
+	w.setIsBlocked(isBlocked)
+	if isBlocked {
+		// If we are blocked, then we want to start the writes.
+		// That function is re-entrant. If we are already writing, then it will just return.
+		w.startWrites()
+	}
+}
+
+// isSemiSyncBlocked checks if the primary is blocked on semi-sync.
+func (w *Watcher) isSemiSyncBlocked(ctx context.Context) (bool, error) {
+	// Get a connection from the pool
+	conn, err := w.appPool.Get(ctx)
+	if err != nil {
+		return false, err
+	}
+	defer conn.Recycle()
+
+	// Execute the query to check if the primary is blocked on semi-sync.
+	res, err := conn.Conn.ExecuteFetch(semiSyncWaitSessionsRead, 1, false)
+	if err != nil {
+		return false, err
+	}
+	// If we have no rows, then the primary doesn't have semi-sync enabled.
+	// It then follows, that the primary isn't blocked :)
+	if len(res.Rows) == 0 {
+		return false, nil
+	}
+
+	// Read the status value and check if it is non-zero.
+	if len(res.Rows) != 1 || len(res.Rows[0]) != 2 {
+		return false, errors.New("unexpected number of rows received")
+	}
+	value, err := res.Rows[0][1].ToInt()
+	return value != 0, err
+}
+
+// waitUntilSemiSyncUnblocked waits until the primary is not blocked
+// on semi-sync.
+func (w *Watcher) waitUntilSemiSyncUnblocked() {
```

Comment:
I don't see where this is used yet. Do we expect multiple watchers? I wouldn't think so. Do we expect a watcher to have more than one goroutine waiting?

Reply:
I made it so that there can be multiple waiters. I still need to hook this up. I'll eventually call this from `DemotePrimary` flow.